adding IPDG support to stress splitting

cpuelz opened the pull request.

cpuelz commented:
@boyceg how can I see the details for the failed checks?  I tried logging into Jenkins and couldn't.

Also, for the heat equation timestepping, I have a separate function "evolveStressNormalization" where I timestep the DOFs.  Should I put this in one of the timestepping functions in IBFEMethod?  Like forwardEulerStep?

cpuelz commented:
@boyceg still haven't looked at the build log, but are the failed checks related to deprecated libMesh code?

cpuelz commented:
Still need to figure out why the initial conditions for the heat equation look weird.

cpuelz commented:
@boyceg I moved the initialization of the heat equation into the computeStressNormalization function.  this seemed to fix things, although I still am not sure why the other way for initializing wasn't working.  I did see that that RHS vectors where different between CG and CG_HEAT at the initial timestep.

cpuelz commented on ibtk/src/lagrangian/FEDataManager.cpp:
```diff
@@ -945,9 +980,209 @@ FEDataManager::prolongData(const int f_data_idx,
+            {
+                F_dof_map_cache.dof_indices(elem, F_dof_indices[i], i);
+            }
+            get_values_for_interpolation(F_node, *F_petsc_vec, F_local_soln, F_dof_indices);
```
@boyceg I'm having an issue with this part of the code and multiple processors.  I get an error thrown from the libMesh::PetscVector function map_global_to_local_index 

No index 192 in ghosted vector.
Vector contains [0,150)
And ghost array {276,277,278,164,279,165,282,283,288,150,289,151,290,291,292,293,156,157}

anything glaringly wrong?

cpuelz replied:
i.e. within get_values_for_interpolation.

boyceg replied:
Were you able to resolve this? What example triggers the problem?

cpuelz replied:
I'm trying to run things again, now in parallel to get some faster simulations, and I am still getting this issue.

cpuelz replied:
if you can, please try running the application my-ex7 in our paper repo in parallel, with eliminate pressure jumps set to true.  I also have a branch off of this repo where I am playing with constructing ghosted vectors in the main routine.

cpuelz replied:
@boyceg re the prolongData functions, I am not sure if I ever fixed this parallel bug.  I ran everything in serial for the paper.

boyceg replied:
OK, well, let's try to get it fixed. :-)

boyceg replied:
I think that the code for getting DOF indices for non-`LAGRANGE` elements may have been broken in parallel. I think that this is fixed in `master`. Are you able to rebase off the current IBAMR master and see if this is fixed?

cpuelz replied:
yeah I can try that.  @drwells, how do I rebase?

boyceg commented:
> On Jun 20, 2018, at 4:47 PM, Charles Puelz <notifications@github.com> wrote:
> 
> @cpuelz commented on this pull request.
> 
> In ibtk/src/lagrangian/FEDataManager.cpp <https://github.com/IBAMR/IBAMR/pull/208#discussion_r196937024>:
> 
> > +            for (unsigned int k = 0; k < n_node; ++k)
> +            {
> +                elem->point(k) = s_node_cache[k];
> +            }
> +
> +            // If there are no intersection points, then continue on to the next
> +            // element.
> +            if (intersection_ref_coords.empty()) continue;
> +
> +            // Evaluate the Lagrangian quantity at the Eulerian grid point and
> +            // update the data on the grid.
> +            for (unsigned int i = 0; i < n_vars; ++i)
> +            {
> +                F_dof_map_cache.dof_indices(elem, F_dof_indices[i], i);
> +            }
> +            get_values_for_interpolation(F_node, *F_petsc_vec, F_local_soln, F_dof_indices);
> @boyceg <https://github.com/boyceg> I'm having an issue with this part of the code and multiple processors. I get an error thrown from the libMesh::PetscVector function map_global_to_local_index
> 
> No index 192 in ghosted vector.
> Vector contains [0,150)
> And ghost array {276,277,278,164,279,165,282,283,288,150,289,151,290,291,292,293,156,157}
> 
> anything glaringly wrong?
> 
Are you spreading using a properly ghosted vector? FEDataManager knows how to construct libMesh vectors with the right ghost information to do Lagrangian-Eulerian interaction.

cpuelz commented:
@boyceg I found a couple small bugs in the prolongData functions (now there are two, one for side and one for cell centered data). This didn't seem to fix the issue.  The code is otherwise the same as the prolongData function in master, which only works for side centered data I think.

As far as usage, I am calling this function in main like:

System& X_system = equation_systems->get_system<System>(IBFEMethod::COORDS_SYSTEM_NAME);
NumericVector<double>* X_vec = X_system.solution.get();
NumericVector<double>* X_ghost_vec = X_system.current_local_solution.get();
X_vec->localize(*X_ghost_vec);
System& Phi_system = equation_systems->get_system<System>(IBFEMethod::PHI_SYSTEM_NAME);
NumericVector<double>* Phi_vec = Phi_system.solution.get();
NumericVector<double>* Phi_ghost_vec = Phi_system.current_local_solution.get();
Phi_vec->localize(*Phi_ghost_vec);
const int phi_idx = ib_method_ops->phi_current_idx;
 fe_data_manager->prolongData(phi_idx,
                                             *Phi_ghost_vec,
                                             *X_ghost_vec,
                                             IBFEMethod::PHI_SYSTEM_NAME,
                                             false,
                                             false,
                                             true,
                                             true);

cpuelz commented:
@boyceg I've tested this with IBFE ex0.  can you have a look and let me know what other things I should change?

cpuelz commented:
@boyceg I have an issue with ambiguity of UniquePtr in FEDataManager for the LLVM c++11 build.

boyceg commented:
> On Jul 17, 2018, at 1:56 PM, Charles Puelz <notifications@github.com> wrote:
> 
> @boyceg <https://github.com/boyceg> I have an issue with ambiguity of UniquePtr in FEDataManager for the LLVM c++11 build.
> 
I am a little surprised. But you can always prepend libMesh::UniquePtr to fix it.

(Incidentally, part of the update to C++11 that @abarret is performing will be to replace libMesh::UniquePtr with std::unique_ptr.)

cpuelz commented:
this should be good to go, but there is sill some weirdness with the LLVM
c++11 build.

On Tue, Jul 17, 2018 at 2:05 PM Boyce Griffith <notifications@github.com>
wrote:

>
>
> > On Jul 17, 2018, at 1:56 PM, Charles Puelz <notifications@github.com>
> wrote:
> >
> > @boyceg <https://github.com/boyceg> I have an issue with ambiguity of
> UniquePtr in FEDataManager for the LLVM c++11 build.
> >
> I am a little surprised. But you can always prepend libMesh::UniquePtr to
> fix it.
>
> (Incidentally, part of the update to C++11 that @abarret is performing
> will be to replace libMesh::UniquePtr with std::unique_ptr.)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/IBAMR/IBAMR/pull/208#issuecomment-405673772>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ALNyy0aKaDwzN9291GxHgfQpftJ0FwUzks5uHidWgaJpZM4S7utL>
> .
>


-- 
Charles Puelz
Postdoc
UNC Math Department
cpuelz.github.io

boyceg commented on examples/IBFE/explicit/ex0/input2d.test:
```diff
@@ -12,7 +12,7 @@ DX0 = L/N                                      // mesh width on coarsest grid le
 MFAC = 4.0                                     // ratio of Lagrangian mesh width to Cartesian mesh width
 ELEM_TYPE = "QUAD9"                            // type of element to use for structure discretization
-CONVERGENCE_STUDY = FALSE                      // indicate whether we are performing a convergence study or not;
+CONVERGENCE_STUDY = TRUE                      // indicate whether we are performing a convergence study or not;
```
Let's not change the behavior of the example input file.

boyceg commented on examples/IBFE/explicit/ex0/input2d.test:
```diff
@@ -41,15 +41,15 @@ PressureInitialConditions {
 DT                         = 0.25*DX           // maximum timestep size
 START_TIME                 = 0.0e0             // initial simulation time
-END_TIME                   = 10*DT               // final simulation time
+END_TIME                   = 0.01              // final simulation time
```
Same as above.

boyceg commented on examples/IBFE/explicit/ex0/input2d.test:
```diff
@@ -66,6 +66,7 @@ OUTPUT_F                   = TRUE
 OUTPUT_OMEGA               = TRUE
 OUTPUT_DIV_U               = TRUE
 ENABLE_LOGGING             = TRUE
+ELIMINATE_PRESSURE_JUMPS   = TRUE
```
Same as above.

boyceg commented on examples/IBFE/explicit/ex0/input2d.test:
```diff
@@ -122,6 +123,7 @@ IBFEMethod {
    use_jump_conditions        = USE_JUMP_CONDITIONS
    use_consistent_mass_matrix = USE_CONSISTENT_MASS_MATRIX
    IB_point_density           = IB_POINT_DENSITY
+   Phi_dt                     = DT
```
The code need to find out the time step size internally and not from the input file --- want to make it easy to use dynamic time step sizes whenever possible.

boyceg replied:
Just wanted to re-up this comment.

cpuelz replied:
It seems like some of the protected functions in ```HierarchyIntegrator``` give me access to some timestep info.  What is the best way to access these in ```IBFEMethod```?

boyceg commented on examples/IBFE/explicit/ex1/input2d.test:
```diff
@@ -20,7 +20,7 @@ SMOOTH_CASE = FALSE
 // solver parameters
 IB_DELTA_FUNCTION          = "IB_4"            // the type of smoothed delta function to use for Lagrangian-Eulerian interaction
-SPLIT_FORCES               = TRUE              // whether to split interior and boundary forces
+SPLIT_FORCES               = FALSE              // whether to split interior and boundary forces
```
Same comment as for the input file above....

boyceg commented on ibtk/include/ibtk/FEDataManager.h:
```diff
@@ -412,12 +412,13 @@ class FEDataManager : public SAMRAI::tbox::Serializable, public SAMRAI::mesh::St
     void prolongData(int f_data_idx,
-                     libMesh::NumericVector<double>& F,
-                     libMesh::NumericVector<double>& X,
+                     libMesh::NumericVector<double>& F_vec,
```
Let's leave these as `F` and `X`.

boyceg commented on ibtk/src/lagrangian/FEDataManager.cpp:
```diff
@@ -869,7 +870,41 @@ FEDataManager::prolongData(const int f_data_idx,
+
+void
+FEDataManager::prolongData_side(const int f_data_idx,
+                                NumericVector<double>& F_vec,
```
`F` and `X`

boyceg commented on ibtk/src/lagrangian/FEDataManager.cpp:
```diff
@@ -895,7 +930,7 @@ FEDataManager::prolongData(const int f_data_idx,
     const DofMap& F_dof_map = F_system.get_dof_map();
     SystemDofMapCache& F_dof_map_cache = *getDofMapCache(system_name);
-    System& X_system = d_es->get_system(system_name);
+    System& X_system = d_es->get_system(COORDINATES_SYSTEM_NAME);
```
No wonder this never seemed to work properly.........

drwells replied:
If this is an actual bug, would you mind moving it to a separate PR so we can get it in asap?

boyceg commented on ibtk/src/lagrangian/FEDataManager.cpp:
```diff
@@ -1077,9 +1112,209 @@ FEDataManager::prolongData(const int f_data_idx,
+{
+    IBTK_TIMER_START(t_prolong_data_cell);
+
+    std::cout << "prolonging cell-centered data from system " << system_name << std::endl;
```
Let's not include these output statements.

boyceg commented on src/IB/IBFEMethod.cpp:
```diff
@@ -224,15 +229,427 @@ get_x_and_FF(libMesh::VectorValue<double>& x,
+// initial condition for the stress normalization function Phi
+// which solves the heat equation.
+Number
+heat_initial_condition(const libMesh::Point& p, const Parameters& parameters, const std::string&, const std::string&)
```
heat--->diffusion ?

boyceg commented on src/IB/IBFEMethod.cpp:
```diff
@@ -224,15 +229,427 @@ get_x_and_FF(libMesh::VectorValue<double>& x,
+Number
+heat_initial_condition(const libMesh::Point& p, const Parameters& parameters, const std::string&, const std::string&)
+{
+    const Real x = p(0);
```
Let's get rid of these unused variables.

boyceg commented on src/IB/IBFEMethod.cpp:
```diff
@@ -224,15 +229,427 @@ get_x_and_FF(libMesh::VectorValue<double>& x,
+                // find id of corresponding neighbor side,
+                // since elements with periodic boundaries may not share the same
+                // side in physical space.
+                int blah = 0;
```
let's not use `foo` and `blah` for variable names...

boyceg commented on src/IB/IBFEMethod.cpp:
```diff
@@ -224,15 +229,427 @@ get_x_and_FF(libMesh::VectorValue<double>& x,
+    DenseMatrix<Number> Ke;
+    std::vector<dof_id_type> dof_indices;
+
+    const Real dt = es.parameters.get<Real>("dt");
```
Need to get the time step size dynamically, or generate an error message if the actual time step size is different from the expected one. In general, we try to make the solvers work "in principle" with variable time step sizes, so that if we ever figure out how to estimate the time step size, we can do it; or so that if we ever get a really efficient implicit solver, it too can dynamically choose its time step sizes.

cpuelz replied:
would there some way to tell when the timestep changes, or should I just track two consecutive timestep sizes and see if they are the same?

cpuelz replied:
in other words, i want to assemble the matrix only when the timestep changes...

cpuelz replied:
@boyceg what do you think?

boyceg replied:
If you can access the member functions for the `HierarchyIntegrator` in this assembly function, then it should be easy to detect.

However, the better thing to do is to make this function always rebuild the full system, but then only call the full assembly function when needed. You can set up a `System` not to reassemble the full system matrix when you call `System::solve()` by setting the member variable `System::assemble_before_solve` to `false`. Then you would just assemble as needed.

You would want to split out the matrix and RHS assembly routines to make this work.

cpuelz replied:
this is how I am doing things now I think.   computeStressNormalization populates the rhs vector, and then I set assemble_before_solve to false.  I just need to fix how I am handling the timestep then?

boyceg replied:
Yes I think so.

I am not sure the best way to communicate the time step size to the solver. I am not crazy about putting extra parameters like the time step size in as a parameter for the full `EquationSystems` object, but that seems to be the usual libMesh way to get parameters into these kinds of routines.

What I do in `FEDataManager` is not to use the libMesh assembly infrastructure, but just to do it myself.

An alternative would be to provide a `libMesh::Assembly` object (see [here](http://libmesh.github.io/doxygen/classlibMesh_1_1System_1_1Assembly.html)) to do the assembly and store the system-specific parameters.

boyceg replied:
If you use `libMesh::Assembly`, you just would have a little class that has an `assemble` member function instead of using a static function to assemble, as in the current implementation.

If you did this, then you could put all of the CG and IPDG parameters directly in the relevant `Assembly` classes. This would cut down a bit on declaring the same variable names over and over. :-)

boyceg commented on src/IB/IBFEMethod.cpp:
```diff
@@ -387,10 +811,42 @@ IBFEMethod::registerStressNormalizationPart(unsigned int part)
+    d_equation_systems[part]->parameters.set<Real>("Phi_diffusion") = d_phi_diffusion;
+
+    // assign function for building Phi linear system.  defaults to CG discretization
+    if (d_phi_solver.compare("CG") == 0)
```
`if (d_phi_solver_type == "CG")`

boyceg commented on src/IB/IBFEMethod.cpp:
```diff
@@ -387,10 +811,42 @@ IBFEMethod::registerStressNormalizationPart(unsigned int part)
+        Phi_system.attach_assemble_function(assemble_cg_poisson);
+        Phi_system.add_variable("Phi_CG", d_phi_fe_order, d_fe_family[part]);
+    }
+    else if (d_phi_solver.compare("IPDG") == 0)
```
`if (d_phi_solver_type == "IPDG")`

boyceg commented on src/IB/IBFEMethod.cpp:
```diff
@@ -387,10 +811,42 @@ IBFEMethod::registerStressNormalizationPart(unsigned int part)
+        Phi_system.attach_assemble_function(assemble_ipdg_poisson);
+        Phi_system.add_variable("Phi_IPDG", d_phi_fe_order, MONOMIAL);
+    }
+    else if (d_phi_solver.compare("CG_HEAT") == 0)
```
`else if (d_phi_solver_type == "CG_DIFFUSION")` ???

boyceg commented on src/IB/IBFEMethod.cpp:
```diff
@@ -1049,10 +1505,20 @@ IBFEMethod::forwardEulerStep(const double current_time, const double new_time)
         d_X_new_vecs[part]->close();
         d_X_half_vecs[part]->close();
 
+        if (d_phi_solver.compare("CG_HEAT") == 0)
```
`if (d_phi_solver_type == "CG_DIFFUSION")`

boyceg commented on src/IB/IBFEMethod.cpp:
```diff
@@ -387,10 +811,42 @@ IBFEMethod::registerStressNormalizationPart(unsigned int part)
+
+    // assign function for building Phi linear system.  defaults to CG discretization
+    if (d_phi_solver.compare("CG") == 0)
+    {
```
Also, each of these strings ("CG", "IPDG", etc) probably should be a constant static member variable, as is done for other "distinguished strings" such as the system names

boyceg commented on src/IB/IBFEMethod.cpp:
```diff
@@ -1192,8 +1658,14 @@ IBFEMethod::spreadForce(const int f_data_idx,
+        PetscVector<double>* Phi_ghost_vec = d_Phi_half_vecs[part];
         X_vec->localize(*X_ghost_vec);
         F_vec->localize(*F_ghost_vec);
+        if(d_is_stress_normalization_part[part])
```
`if (`

boyceg commented on src/IB/IBFEMethod.cpp:
```diff
@@ -1903,7 +2456,14 @@ IBFEMethod::computeStressNormalization(PetscVector<double>& Phi_vec,
                                                            data_time,
                                                            d_lag_surface_force_fcn_data[part].ctx);
-                    Phi -= n * F_s * dA_da;
+                    if (d_scale_phi_by_J)
```
Instead of `scale_phi_by_J`, what about something about whether phi is define with respect to reference or current coordinates?

cpuelz commented:
@boyceg I think I am done addressing your comments.  maybe you can look at the last commit to see if that's close to what you had in mind, or if this should be done in another way?

boyceg commented:
> tested convergence rates for IBFE ex0, with CG, IPDG, and the diffusion equation, with diffusion constant O(1) and O(dx^2).
> 

O(dx) is also interesting, I think?

cpuelz commented:
yeah, I think so.  it looks like the number of linear solver iterations is bounded if diffusion is O(dx).  I guess this is because you are premultiplying the Laplacian with dt also, when solving the heat/diffusion equation?

boyceg commented:
@cpuelz, in the end, do we want to be able to use IPDG for the stress splitting?

cpuelz commented:
@boyceg it could be nice to have this just in case, since it is just an extra assembly function, and it allows us to the p-refinement easily.  by the way, this branch has all the stress splitting stuff, not just IPDG.

boyceg commented:
OK, great. In that case, let's clean it up.

@drwells can you give Charles some pointers on cleaning up history? :-D

cpuelz commented:
can I also set up the "inflating ring" problem as an example in IBFE?  as
far as I can tell, this is new test case with analytic solution.

On Fri, Apr 12, 2019 at 4:49 PM Boyce Griffith <notifications@github.com>
wrote:

> OK, great. In that case, let's clean it up.
>
> @drwells <https://github.com/drwells> can you give Charles some pointers
> on cleaning up history? :-D
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/IBAMR/IBAMR/pull/208#issuecomment-482717968>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ALNyy-H_VorXVQo0M_3IMTn_gOF0VDXeks5vgPE-gaJpZM4S7utL>
> .
>


-- 
Charles Puelz
Courant Institute, NYU
cpuelz.github.io

boyceg commented:
New tests are always great!

Want to make it work in 3D also? :-D

cpuelz commented:
that would be cool :-).  yeah, we can do that.

On Fri, Apr 12, 2019 at 4:55 PM Boyce Griffith <notifications@github.com>
wrote:

> New tests are always great!
>
> Want to make it work in 3D also? :-D
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/IBAMR/IBAMR/pull/208#issuecomment-482720085>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ALNyyxdD7Utb4qxtSEt-q9PKy0xLpw6Xks5vgPLLgaJpZM4S7utL>
> .
>


-- 
Charles Puelz
Courant Institute, NYU
cpuelz.github.io

drwells commented on src/IB/IBFEMethod.cpp:
```diff
@@ -98,14 +99,18 @@
 #include "libmesh/petsc_vector.h"
 #include "libmesh/point.h"
 #include "libmesh/point_locator_base.h"
+#include "libmesh/point_locator_tree.h"
```
Do we need nanoflann for this to work?

drwells commented:
I'm looking at this branch now. I think my main suggestion for history clean up would be to just reorganize everything into a few commits: something like "new feature 1", "new feature 2", "update all examples to use new feature 1", etc. I would proceed by just completely sweeping away the old history with
```
git reset --soft master
```
and then staging and committing chunks related to each feature. Moving chunks around is really hard to do from the command line: do you use something like magit or git-kraken?

drwells commented on ibtk/src/lagrangian/FEDataManager.cpp:
```diff
@@ -1077,9 +1112,207 @@ FEDataManager::prolongData(const int f_data_idx,
+
+    IBTK_TIMER_STOP(t_prolong_data_cell);
+    return;
+} // prolongData_cell
```
This function is nearly identical to the face-centered version. I would very much prefer if we combined them and just had a few if statements to switch between face/cell algorithms inside the element loop.

drwells commented:
If the inflating ring can be run quickly (i.e., its run time is about 30 seconds or less) it would be great to add it to the test suite.

boyceg commented on examples/IBFE/explicit/ex0/example.cpp:
```diff
@@ -534,63 +612,36 @@ bool run_example(int argc, char** argv, std::vector<double>& u_err, std::vector<
                  << "  max-norm: " << hier_cc_data_ops.maxNorm(p_cloned_idx, wgt_cc_idx) << "\n"
                  << "+++++++++++++++++++++++++++++++++++++++++++++++++++\n";
-
+            
```
Let's get rid of this trailing whitespace.

boyceg commented on examples/IBFE/explicit/ex0/input2d.test:
```diff
@@ -190,13 +191,14 @@ Main {
 
 // hierarchy data dump parameters
-   data_dump_interval          = 0
+   data_dump_interval          = 1
```
Don't change this parameter in the default input file.

boyceg commented on examples/IBFE/explicit/ex0/input2d.test:
```diff
@@ -190,13 +191,14 @@ Main {
    timer_dump_interval         = 0
 }
 
+
```
```suggestion
```

boyceg commented on examples/IBFE/explicit/ex1/example.cpp:
```diff
@@ -496,8 +541,17 @@ run_example(int argc, char* argv[])
+            }
         }
-
+        
```
```suggestion
```

boyceg commented on examples/IBFE/explicit/ex1/input2d.test:
```diff
@@ -45,6 +45,8 @@ OUTPUT_F                   = TRUE
 OUTPUT_DIV_U               = TRUE
 ENABLE_LOGGING             = TRUE
+ELIMINATE_PRESSURE_JUMPS   = TRUE
+
```
```suggestion
```

boyceg commented on examples/IBFE/explicit/ex1/input2d.test:
```diff
@@ -101,7 +103,8 @@ IBFEMethod {
    use_consistent_mass_matrix = USE_CONSISTENT_MASS_MATRIX
    IB_point_density           = IB_POINT_DENSITY
-}
+   Phi_dt                     = DT
```
The solvers need to be able to figure out the time step size internally. This should not be an input parameter.

boyceg commented on ibtk/include/ibtk/FEDataManager.h:
```diff
@@ -412,9 +412,10 @@ class FEDataManager : public SAMRAI::tbox::Serializable, public SAMRAI::mesh::St
-     * \brief Prolong a value or a density from the FE mesh to the Cartesian
-     * grid.
-     */
+    * \brief Prolong a value or a density from the FE mesh to the Cartesian
```
```suggestion
     * \brief Prolong a value or a density from the FE mesh to the Cartesian
```

boyceg commented on ibtk/include/ibtk/FEDataManager.h:
```diff
@@ -412,9 +412,10 @@ class FEDataManager : public SAMRAI::tbox::Serializable, public SAMRAI::mesh::St
-     * grid.
-     */
+    * \brief Prolong a value or a density from the FE mesh to the Cartesian
+    * grid.  This function contains a switch to handle different types of
```
```suggestion
     * grid.  This function contains a switch to handle different types of
```

boyceg replied:
(etc.)

boyceg commented on ibtk/src/lagrangian/FEDataManager.cpp:
```diff
@@ -861,15 +862,49 @@ FEDataManager::spread(const int f_data_idx,
 FEDataManager::prolongData(const int f_data_idx,
-                           NumericVector<double>& F_vec,
-                           NumericVector<double>& X_vec,
+                           NumericVector<double>& F,
```
Why are these names changed?

boyceg commented on ibtk/src/lagrangian/FEDataManager.cpp:
```diff
@@ -1703,7 +1965,214 @@ FEDataManager::restrictData(const int f_data_idx,
+} // restrictData_side
+
+void
+FEDataManager::restrictData_cell(const int f_data_idx,
```
Same comment as David's above --- we should try to avoid multiple implementations of nearly identical functionality. Can we have just a single function that can switch between side- and cell-centered representations?

boyceg commented on include/ibamr/IBFEMethod.h:
```diff
@@ -794,9 +805,19 @@ class IBFEMethod : public IBStrategy
+    bool d_phi_current_config;
     bool d_has_stress_normalization_parts;
     std::vector<bool> d_is_stress_normalization_part;
+    double d_dt_previous;
```
This is already provided by the `HierarchyIntegrator` base class via `HierarchyIntegrator::d_dt_previous`

boyceg commented on src/IB/IBFEMethod.cpp:
```diff
@@ -78,6 +78,7 @@
 #include "ibtk/libmesh_utilities.h"
 #include "libmesh/auto_ptr.h"
 #include "libmesh/boundary_info.h"
+#include "libmesh/boundary_info.h"
```
```suggestion
```

boyceg commented on src/IB/IBFEMethod.cpp:
```diff
@@ -226,15 +231,413 @@ get_x_and_FF(libMesh::VectorValue<double>& x,
 
-static const Real PENALTY = 1.e10;
+void
+assemble_cg_diffusion(EquationSystems& es, const std::string& system_name)
```
Possibly there are other implementations that could benefit from these assembly routines. Would it make sense to move them to a separate libMesh utility class (e.g. in IBTK)?

cpuelz replied:
yeah it could be nice to do this.

boyceg commented on src/IB/IBFEMethod.cpp:
```diff
@@ -1758,22 +2262,36 @@ IBFEMethod::computeStressNormalization(PetscVector<double>& Phi_vec,
+    const Real ipdg_beta0 = equation_systems->parameters.get<Real>("ipdg_beta0");
+    const std::string Phi_solver = equation_systems->parameters.get<std::string>("Phi_solver");
+    const Real diffusion = equation_systems->parameters.get<Real>("Phi_diffusion");
+    const Real dt = equation_systems->parameters.get<Real>("dt");
```
We shouldn't read in `dt` this way.

cpuelz replied:
yeah I know... this was a hack.  I'll setup something like you suggested above.

boyceg commented on src/IB/IBFEMethod.cpp:
```diff
@@ -1758,22 +2262,36 @@ IBFEMethod::computeStressNormalization(PetscVector<double>& Phi_vec,
     FEType Phi_fe_type = Phi_dof_map.variable_type(0);
     std::vector<int> Phi_vars(1, 0);
 
+    // things for building RHS of Phi linear system based on poisson solver.
```
These parameter names _might_ be a little clearer in the input files if they were grouped together in their own `Database`, so that you don't have to worry about prefixing them with `cg` or `ipdg`

In the input file, you could have
```
IBFEMethod {
   // other settings...
   StressSplitting {
      penalty = ...
      formulation = ...
   }
}
```
You can access this sub-`Database` via
```
    if (input_db->isDatabase("StressSplitting"))
    {
        Pointer<Database> stress_splitting_db = input_db->getDatabase("StressSplitting");
        // ... etc ...
    }
```

cpuelz replied:
yes, this would be much cleaner!!

boyceg commented on src/IB/IBFEMethod.cpp:
```diff
@@ -1882,7 +2423,14 @@ IBFEMethod::computeStressNormalization(PetscVector<double>& Phi_vec,
-                        Phi += n * ((PP * FF_trans) * n) / J;
+                        if (d_phi_current_config)
+                        {
+                            Phi += n * ((PP * FF_trans) * n) / J;
```
I would suggest adding a variable instead of this `if` statement:
```
                        double volume_scale = (d_phi_current_config ? 1.0 / J : 1.0);
                        Phi += n * ((PP * FF_trans) * n) * volume_scale;
```
(Same comment below.)

boyceg commented on src/IB/IBFEMethod.cpp:
```diff
@@ -1905,7 +2453,14 @@ IBFEMethod::computeStressNormalization(PetscVector<double>& Phi_vec,
                                                            data_time,
                                                            d_lag_surface_force_fcn_data[part].ctx);
-                    Phi -= n * F_s * dA_da;
+                    if (d_phi_current_config)
```
Same comment as above.

boyceg commented on src/IB/IBFEMethod.cpp:
```diff
@@ -1931,29 +2486,80 @@ IBFEMethod::computeStressNormalization(PetscVector<double>& Phi_vec,
                                                               data_time,
                                                               d_lag_surface_pressure_fcn_data[part].ctx);
-                    Phi += P;
+                    if (d_phi_current_config)
```
Same comment as above.

boyceg commented on src/IB/IBFEMethod.cpp:
```diff
@@ -1931,29 +2486,80 @@ IBFEMethod::computeStressNormalization(PetscVector<double>& Phi_vec,
                 for (unsigned int i = 0; i < n_basis; ++i)
                 {
-                    Phi_rhs_e(i) += PENALTY * Phi * phi_face[i][qp] * JxW_face[qp];
+                    if ((solver_flag.compare(CG_PHI_SOLVER_NAME) == 0) || (solver_flag.compare(CG_DIFFUSION_PHI_SOLVER_NAME) == 0))
```
@drwells is there a nicer C++11 way to do this?

boyceg commented on src/IB/IBFEMethod.cpp:
```diff
@@ -1931,29 +2486,80 @@ IBFEMethod::computeStressNormalization(PetscVector<double>& Phi_vec,
     Phi_system.solution->localize(Phi_vec);
-    Phi_dof_map.enforce_constraints_exactly(Phi_system, &Phi_vec);
+
+    if ((solver_flag.compare(CG_PHI_SOLVER_NAME) == 0) || (solver_flag.compare(CG_DIFFUSION_PHI_SOLVER_NAME) == 0))
```
Can you add a brief comment explaining why you do not need to enforce constraints for other solver choices?

boyceg commented on src/IB/IBFEMethod.cpp:
```diff
@@ -2252,15 +2856,24 @@ IBFEMethod::computeInteriorForceDensity(PetscVector<double>& G_vec,
                 // at the quadrature point and add the corresponding forces to
                 // the right-hand-side vector.
-                PP = -J * Phi * FF_inv_trans;
+                if (d_phi_current_config)
```
Same comment as above --- make a variable for the volume scaling.

boyceg commented on src/IB/IBFEMethod.cpp:
```diff
@@ -2708,6 +3348,7 @@ IBFEMethod::spreadTransmissionForceDensity(const int f_data_idx,
     if (!d_split_normal_force && !d_split_tangential_force) return;
 
     // Check to see if we need to integrate the surface forces.
+
```
```suggestion
```

boyceg commented on src/IB/IBFEMethod.cpp:
```diff
@@ -2738,6 +3379,9 @@ IBFEMethod::spreadTransmissionForceDensity(const int f_data_idx,
     const unsigned int dim = mesh.mesh_dimension();
 
     // Extract the FE systems and DOF maps, and setup the FE object.
+    System* Phi_system = Phi_vec ? &equation_systems->get_system(PHI_SYSTEM_NAME) : NULL;
```
`nullptr`

boyceg commented on src/IB/IBFEMethod.cpp:
```diff
@@ -2758,6 +3402,8 @@ IBFEMethod::spreadTransmissionForceDensity(const int f_data_idx,
     fe.evalQuadraturePointsFace();
     fe.evalQuadratureWeightsFace();
     const size_t X_sys_idx = fe.registerInterpolatedSystem(X_system, vars, vars, &X_ghost_vec);
+    const size_t Phi_sys_idx = Phi_vec ? fe.registerInterpolatedSystem(*Phi_system, Phi_vars, no_vars, Phi_vec) :
```
I think we now have an `invalid_index` that you could use here.

boyceg replied:
Whoops, that is for SAMRAI indices, not libMesh.

Is there a libMesh invalid index variable? I am not sure...

boyceg commented on src/IB/IBFEMethod.cpp:
```diff
@@ -2859,6 +3505,24 @@ IBFEMethod::spreadTransmissionForceDensity(const int f_data_idx,
+                        // Compute the value of the first Piola-Kirchhoff stress tensor
+                        // at the quadrature point and add the corresponding forces to
+                        // the right-hand-side vector.
+                        if (d_phi_current_config)
```
Same comment as above --- make a variable for the volume scaling and get rid of the `if`-`else` blocks.

boyceg commented on src/IB/IBFEMethod.cpp:
```diff
@@ -3492,7 +4157,17 @@ IBFEMethod::commonConstructor(const std::string& object_name,
     // Indicate that all of the parts do NOT use stress normalization by default
     // and set some default values.
-    d_epsilon = 0.0;
+    d_phi_epsilon = 0.0;
```
You _might_ want to group these together into `struct`s, to make the name spacing "nicer".

drwells commented:
This branch is pretty stale. @cpuelz what is your plan for finishing this?

cpuelz commented:
> This branch is pretty stale. @cpuelz what is your plan for finishing this?

@drwells thanks for reminding me.  I think the code in this branch has been incorporated in the newer branch ```cpuelz/stress-splitting``` with PR #606.  I can finish working on that, although I need to spend some time figuring out what needs to be done.  If you want to spend some time chatting about that, let me know.

drwells commented:
Lets close this in favor of #606.
